lang/java/reef-runtime-standalone/src/main/java/org/apache/reef/runtime/standalone/driver/NodeListManager.java

```diff
@@ -0,0 +1,304 @@
+  private final double jvmHeapFactor;
+  private final REEFEventHandlers reefEventHandlers;
+  private final String errorHandlerRID;
+  private final String rootFolder;
+  private final String nodeFolder;
+  private final int sshPort = 22;
+
+  @Inject
+  NodeListManager(@Parameter(NodeInfoSet.class) final Set<String> nodeInfoSet,
+                  final ConfigurationSerializer configurationSerializer,
+                  final REEFFileNames fileNames,
+                  @Parameter(JVMHeapSlack.class) final double jvmHeapSlack,
+                  final RemoteManager remoteManager,
+                  final REEFEventHandlers reefEventHandlers,
+                  @Parameter(RootFolder.class) final String rootFolder,
+                  @Parameter(NodeFolder.class) final String nodeFolder) {
+    this.nodeInfoSet = nodeInfoSet;
+    this.configurationSerializer = configurationSerializer;
+    this.fileNames = fileNames;
+    this.jvmHeapFactor = 1.0 - jvmHeapSlack;
+    this.reefEventHandlers = reefEventHandlers;
+    this.errorHandlerRID = remoteManager.getMyIdentifier();
+    this.rootFolder = rootFolder;
+    this.nodeFolder = nodeFolder;
+
+    this.nodeSetIterator = this.nodeInfoSet.iterator();
+
+    remoteManager.registerHandler(ReefServiceProtos.RuntimeErrorProto.class,
+        new EventHandler<RemoteMessage<ReefServiceProtos.RuntimeErrorProto>>() {
+          @Override
+          public void onNext(final RemoteMessage<ReefServiceProtos.RuntimeErrorProto> value) {
+            final FailedRuntime error = new FailedRuntime(value.getMessage());
+            LOG.log(Level.SEVERE, "FailedRuntime: " + error, error.getReason().orElse(null));
+            release(error.getId());
+          }
+        });
+
+    LOG.log(Level.FINEST, "Initialized NodeListManager.");
+  }
+
+  void release(final String containerID) {
+    synchronized (this.containers) {
+      final Container ctr = this.containers.get(containerID);
+      if (null != ctr) {
+        LOG.log(Level.INFO, "Releasing Container with containerId [{0}]", ctr);
+        if (ctr.isRunning()) {
+          ctr.close();
+        }
+        this.containers.remove(ctr.getContainerID());
+      } else {
+        LOG.log(Level.INFO, "Ignoring release request for unknown containerID [{0}]", containerID);
+      }
+    }
+  }
+
+  public void onResourceLaunchRequest(final ResourceLaunchEvent resourceLaunchEvent) {
+    LOG.log(Level.INFO, "NodeListManager:onResourceLaunchRequest");
+
+    final String remoteNode;
+    try {
+      synchronized (this.nodeSetIterator) {
+        remoteNode = this.getNode();
+      }
+    } catch (Exception e) {
+      throw new RuntimeException("Unable to get remote node", e);
+    }
+    final String username = remoteNode.substring(0, remoteNode.indexOf('@'));
+    final String hostname = remoteNode.substring(remoteNode.indexOf('@') + 1);
+    final String userHomeDir = System.getProperty("user.home");
+    final String privatekey = userHomeDir + "/.ssh/id_dsa";
```

Comment:
Any chance that L136-L139 will throw an exception? If so, would there be a more friendly way to tell the user what went wrong?